### PR TITLE
VQ-VAE + GAN for OASIS brain image generation

### PR DESCRIPTION
**Pull request by rwardd**
This pull request contains the source files, results and instructions to achieve Task 8 - Creating a generative model for the OASIS Brain dataset.  The VQ-VAE was successfully implemented with a SSIM > 0.6. A GAN was trained to produce novel quatized representations of input images, which were used as inputs to the VQ-VAE decoder to generate original brain images.  

The files of interest are listed below:
```modules.py```: Contains the Encoder, Decoder and Vector Quantisation layers that make up the VQ-VAE. Also contains a simple GAN for novel image generation.
```predict.py```: Contains functions to generate output results as seen in the ```README.md```
```train.py```: Contains the training algorithms for the VQ-VAE and GAN
```dataset.py```: Contains a custom OASIS dataset and the respective functions to build the train, test and validate data loaders.
```constants.py```: Contains the hyperparameters, filepaths and architecture parameters for training and testing.
```driver.py```: The main entry point to the task. Will train, save the models, then generate novel images of the OASIS dataset.

### CODE REVIEW

**Comment by SiyuLiu0329** [comment]
**This is an initial inspection, no action is required at this point**

Difficulty: Hard

Readme Overall:  Good
- Project Overview: Good
- Model: Good
- Data + Preprocessing: No description of pre-precessing techniques
- Usage: no code usage
- Training / Loss Curve(s): Good
- Result Demonstration: Good
- References: Good

Functionality:
- Reconstruction: Good
- Generation: Good

Code:
- Consistent with Results: Yes
- File Structure: Good
- Commenting: Good
- Commit frequency: Good
- Commit messages: Good

Other Comments:
N/A

**Comment by rwardd** [comment]
Can I have some elaboration on the 'no code usage' please? Lines 65-66 in my README state how to reproduce the results (i.e. usage), and ```driver.py``` contains all necessary code to train the models and generate results. Pre-processing was also described in lines 21-22 of the README.

Thank you.

**Comment by shakes76** [comment]
# Marking
## Good Practice (Design/Commenting, TF/Torch Usage)
Adequate design and implementation
Good spacing and comments
Header blocks
## Recognition Problem 
Solves problem 
Driver Script present
File structure present
Shows Usage & Demo & Visualisation & Data usage
Module present
Commenting
No Data leakage
Difficulty: Hard
## Commit Log
Meaningful commit messages
Progressive commits used
## Documentation
ReadMe good
Model/technical explanation
Good Description and Comments 
Markdown used and PDF submitted
## Pull Request
Successful Pull Request (Working Algorithm Delivered on Time in Correct Branch) 
No Feedback required
Request Description good
